Merge Dev changes into Master

LordOfPolls opened the pull request:
## What type of pull request is this?

<!-- Check whichever applies to your PR -->
- [ ] Non-breaking code change
- [x] Breaking code change
- [ ] Documentation change/addition

## Description
<!-- Clearly and concisely describe what this PR is for, and why you feel it should be merged. -->
Merges Dev Changes into Master

## Changes
<!-- - A bullet pointed list outlining the changes you have made -->
- Fill subcommand docstrings if `sub_cmd_description` isnt set
- Remove ghost subcommand from declarative subcommand pattern
- Utilise ratelimit buckets in http client 
- Handle global ratelimit being exceeded in http client
- Throttle http calls
- Improvements to `guild.prune`
- Log rate limits/remaining calls
- Update channel pin timestamp
- 💥Return default role in `Member.top_role`
- 💥Set default channel position to last 
- 💥Fix Broken guild get/fetch thread
- Allow using event objects as parameter in `listen` decorator
- More import changes!
    - `__all__` is now used
    - Unnecessary imports removed
    - Removes polluting objects from root namespace
- Keep local member model in sync with role changes
- Fix DM commands 
- CI updates

## Checklist

<!-- These are actions you **must** have taken, if you haven't, your PR will be rejected -->
- [x] I've formatted my code with [Black](https://black.readthedocs.io/en/stable/)
- [x] I've ensured my code works on `Python 3.10.x`
- [x] I've tested my code

silasary committed:
fix: Subcommand improvements (#216)

* fix: Remove ghost subcommand from declarative subcommand pattern

* feat: Fill subcommand docstrings if `sub_cmd_description` isn't set

LordOfPolls committed:
feat: improve http requests

* feat: Improve http requests

- Adds more functionality to Route object
- Remove categorically ugly logic for bucket rate limiting
- Reduce attempts from 5 -> 3
- Ensure lock is always released for endpoints

* fix: Ensure correct endpoint used when deleting reactions

* fix: Remove magic numbers

So proxy won't beat me

* fix: use `get_running_loop`

Co-authored-by: Bluenix <bluenixdev@gmail.com>

* fix: Allow error to propagate normally

* feat: add __eq__ to Route

* feat: Return `NotImplemented` on invalid comparison

(proxy asked for this)

Co-authored-by: Bluenix <bluenixdev@gmail.com>

LordOfPolls committed:
feat: Throttle gateway sends to prevent gateway abuse

Hi blue 👋

Basically a lovely hack to the cooldown system that throttles the frequency we send payloads to the gateway. Will be replaced later when i revamp cooldowns. This should help prevent people accidentally spamming the gateway on re-connections

LordOfPolls committed:
fix: Incorrect usage of HTTP ratelimiter

LordOfPolls committed:
fix: Disregard method in rate limit bucket allocation

> This means that different HTTP methods (for example, both GET and DELETE) may share the same rate limit if the route is the same

-https://discord.com/developers/docs/topics/rate-limits

LordOfPolls committed:
feat: Log rate limit limits/remaining

leestarb committed:
fix: Improve guild prune (#218)

- Fix error when .join() got int
- Use to_snowflake()
- Refactor the roles argument
- Adjust return type annotations

leestarb committed:
fix: Update channel pin timestamp (#219)

leestarb committed:
fix💥: Return default role in `Member.top_role` (#222)

* fix!: Return default role in `Member.top_role`

* docs: Remove redundant info

LordOfPolls committed:
feat: Allow using event objects as parameter listen decorator

ie:
`@listen(events.GuildJoin)`

LordOfPolls committed:
refactor: Optimise imports of models

leestarb committed:
docs: Change import for correct type annotation (#223)

leestarb committed:
fix💥: Set default channel position to last (#220)

* fix!: Set default channel position to last

* refactor: Use Absent/MISSING

leestarb committed:
docs: Improve docs (see desc.) (#224)

* docs: Improve docs

- Add missing docs in `Role.members`
- Fix wrong note format in `Role.is_assignable`

* docs: Fix docs in `Guild.members`

* docs: Use Absent & Optional

To be same as in other create channel methods

Co-authored-by: LordOfPolls <22540825+LordOfPolls@users.noreply.github.com>

silasary committed:
fix: Keep local member model in sync when we add or remove roles (#228)

silasary committed:
fix: If a message is from a DM, don't try setting its guild_id (#231)

leestarb committed:
docs: Use Absent & Optional (#225)

To be same as in other create channels

LordOfPolls committed:
feat: Respect http buckets in http client for rate limiting (#221)

* feat: Respect http buckets in http client for rate limiting

* fix: Use `route.endpoint` in bucket creation

This means all requests will now respect discords spec for rate limit buckets

* fix: prevent Runtime errors when lock is released after already having been deferred released

* refactor: Minor variable changes and call reorganising

* docs: Document ratelimiting code

* feat: Use WeakValueDict for locks

* feat: Add ratelimit data to debug cache cmd

* feat: handle global rate limits being exceeded

LordOfPolls committed:
feat 💥: Dunder "all" the files (#229)

Note: this **will** break your imports if youre using eronious imports

* feat: Add `__all__` to snek models

* feat: add `__all__` to discord models

* feat: Add `__all__` to api package

* feat: Add `__all__` to const

* feat: Add `__all__` to errors

* feat: Add `__all__` to client

* feat: Add `__all__` to mixins

* feat: Add `__all__` to utils

* feat: Add `__all__` to debug scale

* feat: Add `__all__` to tasks

* feat: Add `__all__` to paginators

LordOfPolls committed:
fix: Ensure const attributes are available in both scopes

This smelly code will be fixed when imports are converted from wildcard importing

leestarb committed:
chore: Fill in requirements url (#233)

leestarb committed:
perf: Remove unnecessary cache check (#232)

LordOfPolls committed:
style: black format

upstream changes meaning our formatting checks fail 😢

LordOfPolls committed:
ci: Version Bump for codeql

LordOfPolls committed:
ci: CodeQL to check dev branch

LordOfPolls committed:
fix: Remove redundant assignment

LordOfPolls committed:
fix: Silence error when discord doesnt send a MessageType

LordOfPolls committed:
Merge branch 'master' into dev

leestarb committed:
fix💥: Broken guild get/fetch thread (#227)

* fix: Broken guild get/fetch thread

- Remove unnecessary `in` check
- Improve the docs

* fix!: Return only channels/threads

* feat: Use `to_snowflake`

leestarb reviewed (approved):
Works good :thumbsup:

LordOfPolls committed:
fix: Ignore ValueError if role has already been removed from `_role_ids`

LordOfPolls committed:
fix: Correct content-type headers

Fixes files being unable to send

leestarb committed:
docs: Sync type annotation  in listener.py(#237)

LordOfPolls committed:
feat: Add target property to interaction context (#235)

* feat: Resolve message data from interactions

* feat: Add target property to InteractionContext

silasary committed:
docs: Added more type annotations, and corrected a few bad ones

silasary committed:
style: A property's setter should be directly below the getter

silasary committed:
docs: Use correct name for PartialEmoji (see #155)

silasary committed:
fix: `listeners` is defined as a property later in the class, this is confusing and redundant

silasary committed:
ci: Remove ANN201 from the ignore list.

Also adds codes from `flake8-unused-arguments` to the ignore list, as we don't care about it yet (but it's in my devenv)

artem30801 committed:
fix: internal import in slash annotations

AstreaTSS committed:
feat: added import namespace for utils (#242)

zevaryx committed:
fix: Reverse byte order on color (#246)

* fix: Reverse byte order on color

* fix: use r, g, b values instead of getting bytes

zevaryx committed:
fix: Asset url and get broken (#247)

artem30801 committed:
fix: usage of @check() decorator with commands

silasary committed:
docs: Correct type hints for `Scale.add_scale_check`

artem30801 committed:
feat: Add guild property to GuildEvent (#251)

zevaryx committed:
feat: Add get util (#243)

* feat: Add get util

* fix: Change logic to skip items with missing attrs

* feat: Add find_all and get_all utility functions

* fix: Use list literal instead of list()

* fix: List comprehension on find_all for better performance

Co-authored-by: LordOfPolls <22540825+LordOfPolls@users.noreply.github.com>

LordOfPolls committed:
chore: Version bump
